### PR TITLE
Allow MultiAppInterpolationTransfer for block restricted sub-app variables

### LINKED ISSUE #12618: MultiAppInterpolationTransfer segfaults if the variable to be transferred is block restricted
## Rationale

While trying to transfer a block restricted variable from sub to master using `MultiAppInterpolationTransfer`, a segfault occurred. I was able to reproduce it on the `moose/test/tests/transfers/multiapp_interpolation_transfer/fromsub` test.

## Description

Here is the patch to reproduce the error (extension changed from .patch to .txt to please Github):
[issue.txt](https://github.com/idaholab/moose/files/2687762/issue.txt)

In dbg mode, I get:

```
Beginning InterpolationTransfer displaced_source_fromsub
Assertion `comp < this->n_comp(s,var)' failed.
comp = 0
this->n_comp(s,var) = 0


Stack frames: 13
0: 0   libmesh_dbg.0.dylib                 0x0000000104646f30 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 2192
1: 1   libmesh_dbg.0.dylib                 0x0000000104631151 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 369
2: 2   libmoose-dbg.0.dylib                0x0000000100ecb0f3 libMesh::DofObject::dof_number(unsigned int, unsigned int, unsigned int) const + 5523
3: 3   libmoose-dbg.0.dylib                0x0000000102220133 MultiAppInterpolationTransfer::execute() + 18435
4: 4   libmoose-dbg.0.dylib                0x0000000101db1479 FEProblemBase::execMultiAppTransfers(MooseEnumItem, MultiAppTransfer::DIRECTION) + 1481
5: 5   libmoose-dbg.0.dylib                0x0000000101d6cbd3 FEProblemBase::execMultiApps(MooseEnumItem, bool) + 2499
6: 6   libmoose-dbg.0.dylib                0x0000000101157618 Transient::solveStep(double) + 712
7: 7   libmoose-dbg.0.dylib                0x00000001011572fa Transient::takeStep(double) + 634
8: 8   libmoose-dbg.0.dylib                0x0000000101156ae3 Transient::execute() + 211
9: 9   libmoose-dbg.0.dylib                0x000000010250754b MooseApp::executeExecutioner() + 235
10: 10  libmoose-dbg.0.dylib                0x0000000102507e9f MooseApp::run() + 335
11: 11  moose_test-dbg                      0x0000000100002353 main + 275
12: 12  libdyld.dylib                       0x00007fffa5ec0235 start + 1
[0] /Users/labovm/projects/moose/scripts/../libmesh/installed/include/libmesh/dof_object.h, line 839, compiled Dec 17 2018 at 11:09:09
```

## Impact

Fixing this issue would allow to use MultiAppInterpolationTransfer with block restricted variables.

### PR DESCRIPTION
**Pull request by vincentlaboure**
Closes #12618 

I don't know if there is a more efficient fix but this seems to work.

### CODE REVIEW

File: test/tests/transfers/multiapp_interpolation_transfer/fromrestrictedsub_master.i [hunk 1]
```diff
@@ -0,0 +1,93 @@
+###########################################################
+# This is a test of the Transfer System. This test
+# uses the Multiapp System to solve independent problems
+# related geometrically. Solutions are then interpolated
+# and transferred from a non-aligned domain.
+#
+# @Requirement F7.20
+###########################################################
```
Comment (permcody) [inline test/tests/transfers/multiapp_interpolation_transfer/fromrestrictedsub_master.i]: Drop this comment, it's not specifically relevant to your new test. Also, we aren't using this old syntax anymore for the requirements.

**Review by permcody — COMMENTED** [review]
This looks ok - go ahead and fixup your failing test

**Comment by vincentlaboure** [comment]
Comments addressed

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
@vincentlaboure - Your new test seems to break down when running on many processors:
https://civet.inl.gov/job/283081/

**Comment by permcody** [comment]
Yeah - it appears this only works if the destination is LAGRANGE. If you look at that method, it _does_ handle both MONOMIALS and LAGRANGE. It looks like you only put the short circuiting in the LAGRANGE calls when you probably should have put it in both places.

**Comment by vincentlaboure** [comment]
We already have the following call for `!from_nodal` and `FROM_MULTIAPP` (which is what is tested by the test):

```
if (from_element->n_comp(from_sys_num, from_var_num) == 0)
  continue;
```

So I am not sure what else to add to the elemental case. Also, these only protect from accessing zero-size vectors so if one of them was missing I would have expected a segfault (which was what that test produced before that PR).

All I can say at this point is that the nodal transfer seems to pass independently of the number of processes but the elemental transfer differs slightly for `p>4`

Am I missing something?

**Pull request merged**